Make collector compatible with ceph luminous

brtkwr opened the pull request.

Commit:
Make collector compatible with ceph luminous

dougszumski reviewed (commented):
Are we able to come up with something which allows the old behaviour to work, but switches to the new behaviour only for our situation? Ideally, we want to submit this upstream and we don't want to break anyone else's workflow. What do you think?

brtkwr commented:
Okay it should be compatible with earlier versions of ceph now.

Commit:
Add rate to metrics

brtkwr commented:
@dougszumski can I merge this branch?

dougszumski commented on monasca_agent/collector/checks_d/ceph.py:
```diff
@@ -51,6 +52,10 @@ def check(self, instance):
+        match_version = re.search(
+                            _CEPH_VERSION_REGEX,
+                            self._ceph_cmd('version'))
+        self.preluminous = match_version.group(1).split('.') < ['12','2','0']
```
How about using something like the snippet below to better describe what '12','2','0' is, and also make the version parsing more robust?

```
>> from pkg_resources import parse_version
>> _CEPH_LUMINOUS_RELEASE_VERSION = '12.2.0'
>> parse_version('12.2.1.dev') > parse_version(_CEPH_LUMINOUS_RELEASE_VERSION)
>> True
```

dougszumski commented on monasca_agent/collector/checks_d/ceph.py:
```diff
@@ -359,10 +365,16 @@ def _get_stats_metrics(self, ceph_status, ceph_status_plain):
+            ceph_status_health[health_key])
 
-        for s in ceph_status_health['summary']:
+        for s in ceph_status_health.get('summary', []):
```
Has the summary field actually been removed or has it been renamed?

dougszumski commented on monasca_agent/collector/checks_d/ceph.py:
```diff
@@ -416,8 +428,8 @@ def _get_mon_metrics(self, ceph_status):
         mon_metrics = {}
-        for health_service in ceph_status['health']['health'][
-                'health_services']:
+        for health_service in ceph_status.get('health').get('health', {}).get(
```
Is this information available, but in a different place?

brtkwr replied:
Hmm I don't actually know, I was just trying to get it to work at the time without coughing errors iirc.

brtkwr replied:
I suppose i could check the monasca agent log on soft iron ceph cluster

dougszumski reviewed (changes requested):
Thanks @brtknr, this gets things going. It would be nice if we could propose it upstream when you're finished so we can avoid deploying our own fork of the monasca agent.

Please could you update the tests (https://github.com/openstack/monasca-agent/blob/master/tests/checks_d/test_ceph.py). It would be nice to see example output from the new ceph luminous command in there to make it clear whether some output is no longer available, or is available but under a different field.

brtkwr commented:
Okay makes sense to close this then.

oneswig committed:
Probe for Ceph admin client key

During detection, probe for the presence of a Ceph admin client key,
without which the ceph commands invoked in the collector will fail.

oneswig committed:
Use local form of Ceph version command.

Instead of running 'ceph version', run 'ceph --version', which does
not appear to require access to the Ceph admin client key.
Provide (in future) for the Ceph version to be written into config
during setup.

dougszumski reviewed (approved):
Changes look good, still a couple of old comments which could be nice to investigate, but we could merge this now it works and follow up later.

I wonder if having a separate plugin for the various Ceph releases would be simpler to maintain in the longer run, particularly as we don't have tests for the various releases.
